ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs

```diff
@@ -273,6 +278,11 @@ instance ShelleyCompatible proto era => QueryLedger (ShelleyBlock proto era) whe
                 , SL._retiring = Map.restrictKeys (SL._retiring dpsPState) poolIds
                 }
             Nothing -> dpsPState
+        GetPoolDistr mPoolIds ->
+          let poolDistr = SL.calculatePoolDistr . SL._pstakeSet . SL.esSnapshots $ getEpochState st in
```

Comment:
Is this `calculatePoolDistr` perhaps too expensive?

Reply:
It might be, but I don't know that we have any clear requirements on what is too expensive. Note that `SL._pstakeSet` is one of the data structure that we anticipate being moved to disk.

Reply:
This is for `query leadership-schedule`, so I think we have a good amount of headroom -- I wouldn't expect SPOs etc to be spamming this command (though it's easy enough to do that accidentally within a script).

Reply:
Is there a way to make this cheaper if we're only interested in a subset of pool ids?

Reply:
great question! The following might be faster:

Make a new version of `calculatePoolDistr` with signature
```
calculatePoolDistr' :: SnapShot crypto -> Set (KeyHash 'Staking crypto) -> PoolDistr crypto
```
which is identical to `calculatePoolDistr` except that after looking up the credential in the delegation map (see [here](https://github.com/input-output-hk/cardano-ledger/blob/dd4b6e31a05e3c02c6f00f0d53cf20e3b8b74468/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/NewEpoch.hs#L214)), it also looks up the pool ID in the set of pool IDs provided by the user (and discards the result if not found).

Reply:
I'd look at use cases before optimising too much. I think the two likely cases are:

- all pools
- a specific pool (and then repeat for multi-pool operators)

The command is likely to be used occasionally rather than frequently, I expect